mapper/collectd_mapper/src/batcher.rs

```diff
@@ -1,169 +1,171 @@
-        collectd_message: CollectdMessage,
-    ) -> Result<(), DeviceMonitorError> {
-        self.message_grouper.measurement(
-            Some(collectd_message.metric_group_key),
-            collectd_message.metric_key,
-            collectd_message.metric_value,
-        )?;
-
-        Ok(())
-    }
-
-    fn end_batch(self) -> MeasurementGrouper {
-        self.message_grouper
+impl message_batcher::BatchingCriterion<OwnedCollectdMessage> for CollectdTimestampDeltaCriterion {
+    fn belongs_to_batch(
+        &self,
+        message: &OwnedCollectdMessage,
+        message_batch: &message_batcher::MessageBatch<OwnedCollectdMessage>,
+    ) -> bool {
+        let delta = message_batch.first().timestamp() - message.timestamp();
+        delta.abs() <= self.delta
     }
 }
 
 pub struct MessageBatcher {
     sender: UnboundedSender<MeasurementGrouper>,
     mqtt_client: Arc<dyn MqttClient>,
     source_topic_filter: TopicFilter,
-    batching_window: Duration,
+    batcher: message_batcher::MessageBatcher<OwnedCollectdMessage>,
     clock: Arc<dyn Clock>,
 }
 
 impl MessageBatcher {
     pub fn new(
         sender: UnboundedSender<MeasurementGrouper>,
         mqtt_client: Arc<dyn MqttClient>,
-        batching_window: Duration,
+        batching_window: chrono::Duration,
         source_topic_filter: TopicFilter,
         clock: Arc<dyn Clock>,
     ) -> Self {
+        let batcher = message_batcher::MessageBatcher::new(1000, batching_window)
+            .with_batching_criterion(CollectdTimestampDeltaCriterion {
+                delta: batching_window.num_seconds() as f64,
+            });
+
         Self {
             sender,
             mqtt_client,
             source_topic_filter,
-            batching_window,
+            batcher,
             clock,
         }
     }
 
-    pub async fn run(&self) -> Result<(), DeviceMonitorError> {
+    pub async fn run(&mut self) -> Result<(), DeviceMonitorError> {
         let mut messages = self
             .mqtt_client
             .subscribe(self.source_topic_filter.clone())
             .await?;
 
+        let mut outputs = Vec::new();
+
         loop {
-            match self.receive_message(messages.as_mut()).await {
-                Some((message, timestamp)) => {
-                    // Build a message batch until the batching window times out and return the batch
-                    let message_batch_result = self
-                        .build_message_batch_with_timeout(message, timestamp, messages.as_mut())
-                        .await;
-
-                    match message_batch_result {
-                        Ok(message_batch) => {
-                            // Send the current batch to the batch processor
-                            let _ = self.sender.send(message_batch).map_err(|err| {
+            let next_notify_in = self.process_outputs(&mut outputs);
+            self.process_io(messages.as_mut(), next_notify_in, &mut outputs)
+                .await;
```

Comment:
It's puzzling to see the outputs processed before the inputs, *until* one grasps how the former control the timeout of the latter.  This would deserve a comment ;-)